File: Decode Ways.java
public class Solution {
    public int numDecodings(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        if(s.isEmpty())
        {
            return 0;   
        }
        int []dp = new int[2];
        int preDigit = s.charAt(s.length() - 1) - '0';
        dp[1] = 1;
        dp[0] = preDigit > 0? 1:0;
       
        for(int i = s.length() - 2; i > -1; i--)
        {
            int oneDigit = (s.charAt(i) - '0');
            int twoDigit = oneDigit*10 + preDigit;
            if( oneDigit == 0)
            {
                dp[1] = dp[0];
                dp[0] = 0;
                continue;
            }
            int predp = dp[0];
            if( twoDigit > 9 && twoDigit <= 26)
            {
                dp[0] += dp[1];
            }
            preDigit = oneDigit;
            dp[1] = predp;
        }
        
        return dp[0];
    }
}

File: Decode Ways.cpp
class Solution {
public:
    int numDecodings(string s) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        if(s.empty())
            return 0;
        
        vector<int> dp(2, 1);
        int pre = 0;
        int len = s.size();
        int num = 0;
        dp[1] = 1;
        int digit  = s[len-1] - '0';
        
        /*
        for dp question, it is not hard to write equation. 
        However the intial value shoule be think twice. Should we start from the first or last or the second or the second from the last
        We should be careful about that!!!!
        */
        
        
        if(digit==0)
        {
            dp[0] = 0;
        }
        else
        {
            dp[0] = 1; 
        }
       
        pre = digit;
        for(int i = len-2; i > -1; i--)
        {
            digit  = s[i] - '0';
            num = pre+10*digit;
            if(digit==0)
            {
                dp[1] = dp[0];
                dp[0] = 0;
                continue;
               
            }
            
            int preDp = dp[0];
            if(num>=10&&num <=26)
                dp[0] +=dp[1];
            dp[1] = preDp;
            pre = digit;
            
        }
        
        return dp[0];
    }
};